### ISSUE TITLE
`tokenizeLine` is not a function

### ISSUE THREAD

**Issue by Denis-net** [microsoft/vscode-textmate#94]
Hello,
I'm Denis, an Italian developer !
I'm trying to craft an advanced postfix completion helper for vscode :
https://github.com/Denis-net/vscode-definitive-postfix

I wanted to activate my suggestions only when not inside comment blocks (tokens),
I tried regexs but I didn't manage to distinguish strings from comment or line comment to block comment etc.

( for example:
``` javascript
var x=5; // /*
x++;
// */
var aString = "//"; x++;
```
)

I discovered that I need a syntax analyzer/tokenizer and after a long search i came across this intresting project.
But I missed something important, because I can't make it working.
I read about _tmLanguage_ and _plist_, I found them on `vscode.env.appRoot/extensions/*/syntaxes` and on
[textmate repo](https://github.com/github/linguist/blob/master/vendor/README.md).

In `extension.ts`.`loadParser`(:110) method I call the `parseRawGrammar`, then I store the result in my global singleton(`SnippetCollector.Current`) field.

In `completionProvider.ts`.`isInsideComment`(:46) method I tried to use the parser, but It raises an exception :

> "`parser.tokenizeLine` is not a function"

so it didn't create the expected object.

Furthermore, in the `loadGrammar.then()`(extension.ts:137) the `grammar` arg is null, while inside the promise I can load a `rawGrammar` object.

Could you help me ?
Maybe you know what I'm doing wrong,
or if there is another way of doing it.

Thanks

**Comment by msftrncs** [comment]
Have you looked at the `inspect.js` script for an example?   Also, are you looking at the console.log for errors (or other relevant location)?

'tokenizeLine' from grammar.ts is a method from the 'Grammar' class, not a function.

I've never built anything using this repository, but I use the 'inspect.js' script to use this repository to help debug grammars.

**Comment by Denis-net** [comment]
Hey @msftrncs ! 😄👋 
Many Thanks for your help !

I followed the [readme.md on the main page](https://github.com/Microsoft/vscode-textmate), so : 

1. I installed the module through `npm install vscode-textmate`
2. I called `new vsctm.Registry` (same code of the main page, the only diffrent things are grammarPaths and a breakpoint)
3. In the `registry.loadgrammar(langId).then([...])` I store the grammar object (and I use the code for parsing multiple lines).

![loadParser](https://user-images.githubusercontent.com/45769702/57009533-69584f00-6bf7-11e9-94d5-37197ecfc823.png)
I went in debug so I placed breakpoints and I discovered this :
The `parseRawGrammar` method returns me this object (RawGrammar, I guess) 
![rawGrammar](https://user-images.githubusercontent.com/45769702/57009540-72492080-6bf7-11e9-8aae-02be69acdb0e.png)


Calling `c(rawGrammar)` the corresponding `loadGrammar.then` is invoked, 
![then](https://user-images.githubusercontent.com/45769702/57009560-8ee55880-6bf7-11e9-80a7-2a168f4ae482.png)
but the resulting `grammar` is empty.
So I stored the rawGrammar, because is my only available instance, but using this Grammar object when I call the `tokenizeLine` method 

![tokenizeLine](https://user-images.githubusercontent.com/45769702/57009573-a3295580-6bf7-11e9-8851-b550a2e8d990.png)
the only error I get on `debug.console` is this
![error](https://user-images.githubusercontent.com/45769702/57009576-aae8fa00-6bf7-11e9-82d0-1fb60bd55ce5.png)
No other messages from the then or other parts.

Hope images will clarify the issue.

using npm I didn't know about `inspect.js`.
Is it possible that the npm version isn't working ?
So I need to download the repo and use it through (inspect.js)[https://github.com/Microsoft/vscode-textmate/blob/master/scripts/inspect.js] ?
I'll give it a try.
👍

**Comment by Denis-net** [comment]
Intresting :
I discovered that the code used in the main window is incorrect.
So I changed it based on what I understood from `inspect.js` and now my code is:
![newLoader](https://user-images.githubusercontent.com/45769702/57010852-34e89100-6bff-11e9-9f88-5121c027f3fe.png)

In Debug I can finally get a valid Grammar object, but when I try to call `tokenizeLine` method, now I get:
![dllIssue](https://user-images.githubusercontent.com/45769702/57010895-5cd7f480-6bff-11e9-90dc-b5c3ea2f502c.PNG)

I don't know what looking for.
But I'm happy, because I got another error, that means: a little step forward ! 🎊🎉 
Have you an idea about this dll ?

**Comment by msftrncs** [comment]
> "Error: A dynamic link library (DLL) initialization routine failed.\?\c:\Progetti\Visual Studio Code Extensions\vscode-definitive-postfix\node_modules\oniguruma\build\Release\onig_scanner.node at process.module.(anonymous function) [as dlopen] (ELECTRON_ASAR.js:166:20) at 

Have you compiled the Atom Node-Origuruma libraries?   These need to be compiled.  This should have been done as part of the dependencies of vscode-textmate.

**Comment by Denis-net** [comment]
I think so,
Sorry, I never did this things before (new to `dependency Inception`: dependency in dependency 😂😅), Still learning 💪🏻

The strange thing is `npm view oniguruma version` showed oniguruma v7.1.0 package installed (in npm_modules);
I tried to re-install [oniguruma](https://github.com/atom/node-oniguruma), through `npm install oniguruma`,
then I also tried to add all the dependencies in package.json (same of the vscode-textmate's package.json), but without luck...
I'll try to rebuild it, somehow 👍🏻.

Btw : will every vscode instance using the extension have to rebuild oniguruma package for being able to use it ?
Or I'll need to build 2(x64, x86) versions of the package and choose one of them at runtime ?

**Comment by Denis-net** [comment]
I tried to compile it but I had issues with node/electron environment (not understood well)...
anyway.

I found a compiled version in `C:\Users\User\AppData\Local\Programs\Microsoft VS Code\resources\app\node_modules.asar.unpacked\oniguruma\build\Release`
So I copied it into my folder and...
the good news is : It works ✨
the bad news is : don't work as I expected 😔

Using plist:
![parsed](https://user-images.githubusercontent.com/45769702/57164199-a6893080-6df3-11e9-8624-4ed2286a4e24.png)

Using tmLanguage I can get more information:
![tmJsonParser](https://user-images.githubusercontent.com/45769702/57164232-c4569580-6df3-11e9-929f-1fc80406ee00.png)

But what I would really like to have are the built-in scope inspector's data :
With It i could understand exactly where I am (more specifically if user cursor is in a comment)

I think I had misunderstood the `vscode-textmate` plugin.
Or I didn't make it working ...

Anyway Thank you @msftrncs  for the great help !

For patching the doc (Readme.md) what can we do ?
Because isn't specified that developers need the Registry and that the addGrammar().then() will return the real Grammar object.
Not speaking about this oniguruma build dependency.

Or can i just close this issue for posterity ?
